SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp

```diff
@@ -329,29 +330,39 @@ void ASpatialDebugger::DrawDebugLocalPlayer(UCanvas* Canvas)
 	}
 }
 
-const FColor& ASpatialDebugger::GetVirtualWorkerColor(const Worker_EntityId EntityId) const
+FColor ASpatialDebugger::GetVirtualWorkerColor(const Worker_EntityId EntityId) const
 {
 	check(NetDriver != nullptr && !NetDriver->IsServer());
-
 	const AuthorityIntent* AuthorityIntentComponent = NetDriver->StaticComponentView->GetComponentData<AuthorityIntent>(EntityId);
-	const int32 VirtualWorkerId = (AuthorityIntentComponent != nullptr) ? AuthorityIntentComponent->VirtualWorkerId : SpatialConstants::INVALID_VIRTUAL_WORKER_ID;
-
-	if (VirtualWorkerId != SpatialConstants::INVALID_VIRTUAL_WORKER_ID &&
-		VirtualWorkerId < ServerTintColors.Num())
-	{
-		return ServerTintColors[VirtualWorkerId];
-	}
-	else
-	{
+	check(AuthorityIntentComponent != nullptr);
+	const int32 VirtualWorkerId = AuthorityIntentComponent->VirtualWorkerId;
+	const FString* PhysicalWorkerId = NetDriver->VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(VirtualWorkerId);
+	if (PhysicalWorkerId == nullptr) {
+		// This can happen if the client hasn't yet received the VirtualWorkerTranslator mapping
 		return InvalidServerTintColor;
 	}
+	return SpatialGDK::GetColourForWorkerId(*PhysicalWorkerId);
 }
 
-// TODO: Implement once this functionality is available https://improbableio.atlassian.net/browse/UNR-2362.
-const FColor& ASpatialDebugger::GetServerWorkerColor(const Worker_EntityId EntityId) const
+FColor ASpatialDebugger::GetServerWorkerColor(const Worker_EntityId EntityId) const
 {
 	check(NetDriver != nullptr && !NetDriver->IsServer());
-	return InvalidServerTintColor;
+
+	const SpatialGDK::EntityAcl* AclData = NetDriver->StaticComponentView->GetComponentData<SpatialGDK::EntityAcl>(EntityId);
+	const WorkerRequirementSet* WriteAcl = AclData->ComponentWriteAcl.Find(SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID);
+
+	check(WriteAcl != nullptr);
+	check(WriteAcl->Num() == 1);
+	check((*WriteAcl)[0].Num() == 1);
+
+	const FString& AuthoritativeACLEntry = (*WriteAcl)[0][0];
```

Comment:
I imagined there would be a helper for this (`const FString& WorkerRequirementSet::Attribute()`), isn't it common to need the attribute string from a requirement set?

Reply:
Discussed offline, in fact it makes more sense to extract the logic for asking what is the name of the server worker which has authority for a given entity id out of here because it has nothing to do with colors or the debugger and it belongs in the core of the GDK. Similarly for asking about the name of the server worker which has authority intent for a given entity id.

Reply:
Yep 👍 as above, this should hopefully be removable with turning this debugger logic to be an actor component